### PR TITLE
Add binary metadata support for cronet transport

### PR DESCRIPTION
**Pull request by muxi**

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: src/core/ext/transport/cronet/transport/cronet_transport.cc [hunk 1]
```diff
@@ -624,14 +653,21 @@ static void on_response_trailers_received(
   for (size_t i = 0; i < trailers->count; i++) {
     CRONET_LOG(GPR_DEBUG, "trailer key=%s, value=%s", trailers->headers[i].key,
                trailers->headers[i].value);
+    grpc_slice key = grpc_slice_intern(
+        grpc_slice_from_static_string(trailers->headers[i].key));
+    grpc_slice value;
+    if (grpc_is_binary_header(key)) {
```
Comment (makdharma) [inline src/core/ext/transport/cronet/transport/cronet_transport.cc]: This code seems duplicated with header processing. Create a separate function and use in both places?

File: src/core/ext/transport/cronet/transport/cronet_transport.cc [hunk 2]
```diff
@@ -214,6 +216,26 @@ void grpc_cronet_stream_unref(stream_obj* s) { grpc_stream_unref(s->refcount); }
 
 static enum e_op_result execute_stream_op(struct op_and_state* oas);
 
+static const size_t tail_xtra[4] = {0, 0, 1, 2};
```
Comment (makdharma) [inline src/core/ext/transport/cronet/transport/cronet_transport.cc]: Is this code (infering length) copied from chttps? Any way to avoid duplication?

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/ext/transport/cronet/transport/cronet_transport.cc (hunk 2 above)
Comment (muxi) [reply src/core/ext/transport/cronet/transport/cronet_transport.cc]: No, chttp does not do this explicitly. However I do think this code should be in `bin_decoder`.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                              FILE SIZE
 ++++++++++++++ GROWING                                                ++++++++++++++
  +0.0%    +224 [None]                                                    +728  +0.0%
  +7.2%    +176 src/core/ext/transport/chttp2/transport/bin_decoder.cc    +176  +7.2%
      [NEW]    +173 grpc_base64_infer_length_after_decode                     +173  [NEW]
       +27%      +3 [Unmapped]                                                  +3   +27%

  +0.0%    +400 TOTAL                                                     +904  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                              FILE SIZE
 ++++++++++++++ GROWING                                                ++++++++++++++
  +0.0%    +288 [None]                                                    +832  +0.0%
  +9.9%    +240 src/core/ext/transport/chttp2/transport/bin_decoder.cc    +240  +9.9%
      [NEW]    +238 grpc_chttp2_base64_infer_length_after_decode              +238  [NEW]
       +18%      +2 [Unmapped]                                                  +2   +18%

  +0.0%    +528 TOTAL                                                  +1.05Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by muxi** [comment]
Failed test is test infrastructure problem

**Pull request merged**